### PR TITLE
Docs fixes for action invocations

### PR DESCRIPTION
**Pull request by mgencur**
More accurate description of blocking calls and the situation when _application error_ is returned upon invoking an action.

## Related issue and scope

- [ ] I opened an issue to propose and discuss this change (#????)

## My changes affect the following components

- [ ] API
- [ ] Controller
- [ ] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [ ] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [ ] Tests
- [ ] Deployment
- [ ] CLI
- [ ] General tooling
- [x] Documentation

## Types of changes

- [x] Bug fix (generally a non-breaking change which closes an issue).
- [ ] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:

- [x] I signed an [Apache CLA](https://github.com/apache/incubator-openwhisk/blob/master/CONTRIBUTING.md).
- [x] I reviewed the [style guides](https://github.com/apache/incubator-openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [ ] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [x] I updated the documentation where necessary.

### CODE REVIEW

File: docs/actions.md [hunk 1]
```diff
@@ -216,10 +216,11 @@ ok: invoked /whisk.system/samples/greeting with id 5975c24de0114ef2b5c24de0118ef
 A blocking invocation request will _wait_ for the activation result to be available. The wait period
 is the lesser of 60 seconds or the action's configured
 [time limit](reference.md#per-action-timeout-ms-default-60s).
-The result of the activation is returned if it is available within the wait period.
-Otherwise, the activation continues processing in the system and an activation ID is returned
-so that one may check for the result later, as with non-blocking requests
-(see [here](#watching-action-output) for tips on monitoring activations).
+The result of the action is successfully returned if it is available within the wait period.
+Otherwise, the result is an error caused by the action timeout and the response status is _application error_.
```
Comment (rabbah) [inline docs/actions.md]: This is not quite true. The API response is 202 (accepted) and an activation id, not an application error (which would imply the activation completed) if the blocking period (60s) expires before a result is ready.

File: docs/actions.md [hunk 2]
```diff
@@ -233,7 +234,10 @@ Each action invocation results in an activation record which contains the follow
 - `response`: A dictionary that defines the following keys
   - `status`: The activation result, which might be one of the following values:
     - *"success"*: the action invocation completed successfully.
-    - *"application error"*: the action invocation was successful, but the action returned an error value on purpose, for instance because a precondition on the arguments was not met.
+    - *"application error"*: the action was invoked, but returned an error value on purpose, for instance because a precondition on the arguments was not met. This status code is also returned under specific conditions such as:
+      - wrong docker container name
+      - initialization of a function failed
+      - action timed out
```
Comment (rabbah) [inline docs/actions.md]: some of these are about to change; see #3941.

File: docs/actions.md [hunk 3]
```diff
@@ -340,7 +344,7 @@ is skipped if an action is dispatched to a previously initialized container ---
 You can tell if an [invocation was a warm activation or a cold one requiring initialization](annotations.md#annotations-specific-to-activations)
 by inspecting the activation record.
 - An action runs for a bounded amount of time. This limit can be configured per action, and applies to both the
-initialization and the execution separately.
+initialization and the execution separately. If the action time limit is exceeded the activation's response status is _application error_.
```
Comment (rabbah) [inline docs/actions.md]: This changes in #3941 - @tysonnorris @markusthoemmes please confirm.

Comment (mgencur) [reply docs/actions.md]: I think it only changes for the init phase.

File: docs/actions.md (hunk 1 above)
Comment (mgencur) [reply docs/actions.md]: This is about blocking invocation. Here's my test:
`wsk action create sleep tests/dat/actions/sleep.js -t 100`
`wsk action invoke sleep --param sleepTimeInMs 1100 --blocking`
The response code is 502 and response itself is:
`ok: invoked /guest/sleep with id 75043a3d2eec4556843a3d2eec855620
{
    "activationId": "75043a3d2eec4556843a3d2eec855620",
    "annotations": [
        {
            "key": "path",
            "value": "guest/sleep"
        },
        {
            "key": "waitTime",
            "value": 1014
        },
        {
            "key": "kind",
            "value": "nodejs:6"
        },
        {
            "key": "limits",
            "value": {
                "logs": 10,
                "memory": 256,
                "timeout": 100
            }
        },
        {
            "key": "initTime",
            "value": 152
        }
    ],
    "duration": 255,
    "end": 1533718832936,
    "logs": [],
    "name": "sleep",
    "namespace": "guest",
    "publish": false,
    "response": {
        "result": {
            "error": "The action exceeded its time limits of 100 milliseconds."
        },
        "status": "application error",
        "success": false
    },
    "start": 1533718832681,
    "subject": "guest",
    "version": "0.0.1"
}
`

File: docs/actions.md (hunk 2 above)
Comment (mgencur) [reply docs/actions.md]: OK. Will change it.

File: docs/actions.md (hunk 1 above)
Comment (markusthoemmes) [reply docs/actions.md]: Ah, the text reads as if you're referring to the blocking wait period. We wait for 60s at most for a blocking invocation to finish, after that we fall back to a 202 (actions can be configured to run longer than 60s). That's what @rabbah is referring to.

Comment (mgencur) [reply docs/actions.md]: I see. Let me rephrase it then.

**Comment by mgencur** [comment]
Hey @rabbah @markusthoemmes , what do you think now? I've fixed the description a little bit in separate commits.

File: docs/actions.md [hunk 4]
```diff
@@ -216,10 +216,17 @@ ok: invoked /whisk.system/samples/greeting with id 5975c24de0114ef2b5c24de0118ef
 A blocking invocation request will _wait_ for the activation result to be available. The wait period
 is the lesser of 60 seconds or the action's configured
 [time limit](reference.md#per-action-timeout-ms-default-60s).
-The result of the activation is returned if it is available within the wait period.
+
+The result of the activation is returned if it is available within the blocking wait period (this wait period is by default 60 seconds as well).
```
Comment (rabbah) [inline docs/actions.md]: does it make sense to fold this line into the sentence on 216 - a bit redundant IMO.

File: docs/actions.md [hunk 5]
```diff
@@ -216,10 +216,17 @@ ok: invoked /whisk.system/samples/greeting with id 5975c24de0114ef2b5c24de0118ef
...
 Otherwise, the activation continues processing in the system and an activation ID is returned
 so that one may check for the result later, as with non-blocking requests
 (see [here](#watching-action-output) for tips on monitoring activations).
+If execution of the action exceeds the action's configured time limit within the blocking wait period
```
Comment (rabbah) [inline docs/actions.md]: If this is strictly intended to explain the activation's response status, then I suggest dropping the sentence on line 224 because what you write next is true for _all_ activations that exceed their configured duration.

File: docs/actions.md [hunk 6]
```diff
@@ -216,10 +216,17 @@ ok: invoked /whisk.system/samples/greeting with id 5975c24de0114ef2b5c24de0118ef
...
+the result of the invocation is an error caused by the action timeout and the response status is _application error_.
```
Comment (rabbah) [inline docs/actions.md]: I don't think we've explained application error or response status at this point; this happens in the next section (understanding the activation record). How about this:

When an action exceeds its configured time lime, the activation record will indicate this error. See [understanding the activation record] for more details.

or something to that effect.

File: docs/actions.md [hunk 7]
```diff
@@ -233,8 +240,11 @@ Each action invocation results in an activation record which contains the follow
 - `response`: A dictionary that defines the following keys
   - `status`: The activation result, which might be one of the following values:
     - *"success"*: the action invocation completed successfully.
-    - *"application error"*: the action invocation was successful, but the action returned an error value on purpose, for instance because a precondition on the arguments was not met.
-    - *"action developer error"*: the action was invoked, but it completed abnormally, for instance the action did not detect an exception, or a syntax error existed.
+    - *"application error"*: the action was invoked, but returned an error value on purpose, for instance because a precondition on the arguments was not met. This status code is also returned under specific conditions such as:
+      - action timed out
```
Comment (rabbah) [inline docs/actions.md]: since there's only one other condition, how about:

This status code is also set when the action exceeds its configured time limit.

File: docs/actions.md [hunk 8]
```diff
@@ -233,8 +240,11 @@ Each action invocation results in an activation record which contains the follow
...
+    - *"action developer error"*: the action was invoked, but it completed abnormally, for instance the action did not detect an exception, or a syntax error existed. This status code is also returned under specific conditions such as:
+      - wrong docker container name
+      - initialization of a function failed
```
Comment (rabbah) [inline docs/actions.md]: - the function failed to initialize for any reason (including time out)
      - a Docker action specified the wrong docker container name
      - a Docker action that does not properly implement the expected runtime protocol (link to actions-new.md)?

File: docs/actions.md (hunk 3 above)
Comment (rabbah) [reply docs/actions.md]: I suggest:

If the action time limit is exceeded during initialization, the activation's response status is ... otherwise it is ...

**Comment by rabbah** [comment]
Thanks for improving the docs - this is important and appreciated.

File: docs/actions.md (hunk 4 above)
Comment (mgencur) [reply docs/actions.md]: AFAIK, the action default timeout is 60 seconds, the blocking wait timeout is by default 60 as well. It could certainly be worded differently but I think it's good as it is.

File: docs/actions.md (hunk 5 above)
Comment (mgencur) [reply docs/actions.md]: Unless the configured time limit is longer than 60 seconds and the blocking wait time is the default. In this case it returns 202 as you said previously? Curious...

File: docs/actions.md (hunk 4 above)
Comment (rabbah) [reply docs/actions.md]: just seems redundant as line 217 already says 60 seconds (you can add this is the default there) and strictly not correct since an action with a timeout less than 60s for an action, it's a little confusing as worded.

File: docs/actions.md (hunk 6 above)
Comment (mgencur) [reply docs/actions.md]: Could be.

File: docs/actions.md (hunk 7 above)
Comment (mgencur) [reply docs/actions.md]: ok

File: docs/actions.md (hunk 8 above)
Comment (mgencur) [reply docs/actions.md]: ok

File: docs/actions.md (hunk 5 above)
Comment (rabbah) [reply docs/actions.md]: There are two things at play:
1. the HTTP response to the client: if the action has not finished in 60 seconds, the status code is 202 and an activation id is returned
2. in all cases (unless the system fails to generate one), the activation record for an action is created and recorded, and if the action timed out, the response status will be application error (for run) or developer error (for init).

Is that clearer?

**Comment by mgencur** [comment]
hey @rabbah, wdyt now?

**Comment by rabbah** [comment]
LGTM now - I’ll leave it to Markus to clarify the init vs run timeout.

**Comment by markusthoemmes** [comment]
@rabbah Clarified in the other PR, I don't see a reason not to make them both behave the same. One could argue that an action timeout is a form of an uncaught exception (as the action is aware of its own deadline). Making them both non-application errors makes sense to me.

**Comment by mgencur** [comment]
@rabbah @markusthoemmes alright, thanks. So let's wait with this PR until the other one is completely sorted out. Then I can make a little change to this one if necessary.

**Comment by mgencur** [comment]
hey, I've updated the description of errors in the event of timeouts, after https://github.com/apache/incubator-openwhisk/pull/3941 was integrated.
